sphinx_immaterial/graphviz.py

```diff
@@ -0,0 +1,509 @@
+      plugins are loaded.  Instead, plugins can only be loaded through a config
+      file.
+
+    Therefore, to ensure LibGD is used for fonts, it is necessary to locate the
+    original config file, parse it and remove the section that loads Pango, if
+    present, and write it to a new temporary directory.
+    """
+
+    orig_config_path: str
+    """Path to original config file."""
+
+    new_config: bytes
+    """New config content with pango excluded."""
+
+
+def get_adjusted_graphviz_config(
+    app: sphinx.application.Sphinx, dot_command: str
+) -> Optional[GraphvizConfigInfo]:
+    """Returns the graphviz configuration info for a given `dot_command`.
+
+    The returned config file must be written to a temporary directory with a
+    symlink "plugins" to the `orig_config_path`.
+    """
+    key = "_sphinx_immaterial_graphviz_adjusted_configs"
+    configs = getattr(app, key, None)
+    if configs is None:
+        configs = {}
+        setattr(app, key, configs)
+    config = configs.get(dot_command, False)
+    if config is False:
+        config = _make_adjusted_graphviz_config(app, dot_command)
+        configs[dot_command] = config
+    return config
+
+
+def _get_orig_config_path(dot_command: str) -> Optional[str]:
+    result = subprocess.run(
+        [dot_command, "-v"], input="", text=True, capture_output=True, check=True
+    )
+
+    m = re.search(
+        r"^The plugin configuration file:\s+(.*)$\s+was successfully loaded\.$",
+        result.stderr,
+        re.MULTILINE,
+    )
+    if m is None:
+        logger.error(
+            "Failed to determine graphviz config path from stderr: %r", result.stderr
+        )
+        return None
+
+    return m.group(1)
+
+
+def _make_adjusted_graphviz_config(
+    app: sphinx.application.Sphinx, dot_command: str
+) -> Optional[GraphvizConfigInfo]:
+    """Determines the graphviz libdir and generates an adjusted config.
+
+    This is called by `get_adjusted_graphviz_config`.
+    """
+
+    orig_config_path = _get_orig_config_path(dot_command)
+    if orig_config_path is None:
+        return None
+
+    config_content = pathlib.Path(orig_config_path).read_bytes()
+
+    # Strip comments
+    config_content = re.sub(b"#[^\n]*", b"", config_content)
+
+    new_config = io.BytesIO()
+
+    prev_index = 0
+
+    def parse_error():
+        logger.error(
+            "Failed to parse graphviz config file %r, starting at: %r",
+            orig_config_path,
+            config_content[prev_index:],
+        )
+
+    found_gd = False
+
+    # Match plugins
+    for m in re.finditer(
+        rb"\s*([^\s{}]+)\s+([^\s{}]+)\s*(\{\s*(?:[^{}\s]+\s*\{[^{}]*\}\s*)*\s*\})\s*",
+        config_content,
+    ):
+        if m.start() != prev_index:
+            parse_error()
+            return None
+        prev_index = m.end()
+        plugin_path = m.group(1)
+        plugin_name = m.group(2)
+        if plugin_name == b"cairo":
+            continue
```

Comment:
What happened to excluding pango? I thought that was a problematic default as well.

Reply:
The path to the plugin includes "pango" but in the config the plugin name is called "cairo".  I think the plugin makes use of both of those libraries.  In any case that line was stale --- instead it now just excludes the `textlayout` functionality from everything except LibGD.  On Windows there is a GDI+ textlayout plugin that we also have to exclude.